analyzers/src/SonarAnalyzer.Common/Rules/DoNotUseDateTimeNowBase.cs

```diff
@@ -0,0 +1,60 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class DoNotUseDateTimeNowBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S6563";
+
+    private static readonly MemberDescriptor DateTimeNow = new(KnownType.System_DateTime, nameof(DateTime.Now));
+    private static readonly MemberDescriptor DateTimeToday = new(KnownType.System_DateTime, nameof(DateTime.Today));
+    private static readonly MemberDescriptor DateTimeOffsetNow = new(KnownType.System_DateTimeOffset, nameof(DateTimeOffset.Now));
```

Comment:
`DateTimeOffset.Now` contains the offset information so local time is defined with a reference to UTC, we shouldn't raise for that

Reply:
`DateTimeOffset` contains the offset information, but when it's converted to DateTime, that information will no longer be stored. The following two `DateTime` values will be different (unless you're in GMT 0 timezone):

```cs
Console.WriteLine(DateTimeOffset.Now.DateTime);
Console.WriteLine(DateTimeOffset.UtcNow.DateTime);
```

[I've added checks for scenarios](https://github.com/SonarSource/sonar-dotnet/blob/c2bcbc719f3b41ace674914026b5fab711a3e5b7/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotUseDateTimeNow.cs#L24) like `DateTimeOffset.Now.DateTime`, as it's the same as `DateTime.Now`.
So I only use this property to check if it's later converted to `DateTime`, not by itself.
The same goes for the `Date` property of `DateTimeOffset`.

Reply:
So, I think targetting specifically `DateTimeOffset.Now.DateTime` is unnecessary (I found 81 results with 2 minutes search in [SourceGraph](https://sourcegraph.com/search?q=context:global+DateTimeOffset.Now.DateTime+file:%5C.cs%24++count:all+select:file+&patternType=standard&sm=0&groupBy=repo)). I mean it's kind of an overkill :D. Having said that, `DateTimeOffset.Now.Date` is the equivalent of DateOnly basically, you're erasing the time and IMO developers are doing it consciously, resulting in FPs most of the time. Anyway, this is my opinion and I don't have facts supporting these thoughts so I'm fine keeping it as is if you don't agree with what I wrote.

Reply:
It's not a common scenario, I agree (although 81 result isn't bad at all). I still added it because:
- one of the recommendations for fixing the issue is to use DateTimeOffset.Now. And one way to misuse this fix is to turn it right back to DateTime with the Date property. So it feels right to keep it.
- it's not a huge effort, 3-4 lines of extra code.